### PR TITLE
add innerTube API, types and tests

### LINKED ISSUE #497: InnerTube duration fetcher
to bypass YouTube rate limits, implement an innertube API endpoint for faster responses and fallback on newleaf if not possible

### PR DESCRIPTION
**Pull request by mchangrh**
closes #497

- [x] emulate entire ApiVideoData for compatibility with newleaf API

### CODE REVIEW

File: src/types/innerTubeApi.model.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+export interface innerTubeVideoDetails {
+    "videoId": string,
+    "title": string,
+    "lengthSeconds": string, // yes, don't ask.
```
Comment (ajayyy) [inline src/types/innerTubeApi.model.ts:4]: :smile:

File: src/utils/getVideoDetails.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+export function getVideoDetails(videoId: string, ignoreCache = false): Promise<videoDetails> {
+    if (!config.newLeafURLs) {
+        return getPlayerData(videoId)
+            .then(data => convertFromInnerTube(data));
+    }
+    return Promise.any([
+        newLeafWrapper(videoId, ignoreCache)
+            .then(videoData => convertFromNewLeaf(videoData)),
+        getPlayerData(videoId)
```
Comment (ajayyy) [inline src/utils/getVideoDetails.ts:54]: Should getPlayerData have a cache too? Right now it always makes both calls, so even if newleaf has it cached, it is still making an innertube call

File: src/utils/innerTubeAPI.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        },
+        videoId: videoID
+    };
+    const result = await axios.post(url, data, {
+        timeout: 3500
+    });
+    if (result.status === 200) {
+        return result.data.videoDetails;
+    } else {
+        return Promise.reject(result.status);
```
Comment (ajayyy) [inline src/utils/innerTubeAPI.ts:22]: For my future review, this is fine because `getVideoDetails` catches it

File: test/cases/innerTubeApi.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { config } from "../../src/config";
+import assert from "assert";
+// import { innerTubeVideoDetails } from "../../src/types/innerTubeApi.model";
```
Comment (ajayyy) [inline test/cases/innerTubeApi.ts:3]: delete?

File: test/cases/innerTubeApi.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { YouTubeAPI } from "../../src/utils/youtubeApi";
+import * as innerTube from "../../src/utils/innerTubeAPI";
+import { partialDeepEquals } from "../utils/partialDeepEquals";
+import { getVideoDetails } from "../../src/utils/getVideoDetails";
+
+const videoID = "dQw4w9WgXcQ";
+const expectedInnerTube = { // partial type of innerTubeVideoDetails
+    videoId: videoID,
+    title: "Rick Astley - Never Gonna Give You Up (Official Music Video)",
```
Comment (ajayyy) [inline test/cases/innerTubeApi.ts:12]: I know it's fun, but maybe a non RIAA test is a better idea

**Review by ajayyy — COMMENTED** [review]
everything else looks good

File: src/utils/getVideoDetails.ts (hunk 2 above)
Comment (mchangrh) [reply src/utils/getVideoDetails.ts:54]: is there a need for innerTube cache? It shouldn't be ratelimited in any capacity, which is what newleaf was for

**Pull request merged**